Content element comments

underscope opened the pull request:
This PR:
- Decouples `Discussion` component
- Enables discussion on `Content Elements`

UI example screenshot:

![Screenshot 2020-11-25 at 13 02 15](https://user-images.githubusercontent.com/32763694/100225411-75fd7000-2f1e-11eb-9680-d8949ecabfe8.png)

underscope committed:
Decoupled Discussion component

- WIP

underscope committed:
Added content element id relationship to comment

underscope committed:
:wrench:

lukasaric committed:
Merge branch 'develop' into feature/ce-comments

lukasaric committed:
Updated migration timestamp

lukasaric committed:
Initial code preparation cleanup :lipstick:

lukasaric committed:
Infer discussion wrapper :rocket:

- Reused common prev sidebar discussion wrapper as
infered component for CE & Activity

lukasaric committed:
Cleanup :surfer:

- Optimized getComments getter
- Create method polish

lukasaric committed:
Attrs cleanup

lukasaric committed:
Merge branch 'develop' into feature/ce-comments

lukasaric committed:
Implemented basic UI :art:

lukasaric committed:
Put CE discussions as part of a menu

lukasaric committed:
Added basic CE discussion notifications

lukasaric committed:
Enabled flaging unseen comments feature

lukasaric committed:
Content element watchers cleanup :nail_care:

lukasaric committed:
Removed leftovers & CE component cleanup

lukasaric committed:
Infered editor discussion component :rocket:

lukasaric committed:
:wrench:

lukasaric committed:
UI tweaks & cleanup

lukasaric committed:
Adjusted marking unseen comments

lukasaric committed:
Decoupled seenBy state

lukasaric committed:
Getters improvement :rocket:

lukasaric committed:
Removed unecessary extra wrapper component

lukasaric committed:
Cleanup & minor watcher fix :surfer:

lukasaric committed:
Minor condition enhancement :eyes:

lukasaric committed:
Naming polish & common discussion wrapper cleanup

lukasaric committed:
Cleanup :nail_care:

lukasaric committed:
Naming polish :eyes:

lukasaric committed:
Comment creation & getter cleanup :lipstick:

lukasaric committed:
Merge branch 'develop' into feature/ce-comments

lukasaric committed:
Added menu transition & reduced width

lukasaric committed:
Attrs Cleanup :lipstick:

lukasaric committed:
Resized CE discussion btn

lukasaric committed:
Minor badge & btn cleanup

lukasaric committed:
Btn position UI improvement :eyes:

lukasaric committed:
:wrench:

lukasaric committed:
Discussion btn UI tweaks :art:

lukasaric committed:
Notification badge UI tweaks :rocket:

lukasaric committed:
Conditionally change icon & badge within btn

lukasaric committed:
Updated vuex persist

underscope commented on server/shared/database/migrations/20201113162457-add_element_id_to_comment.js:
```diff
@@ -0,0 +1,12 @@
+'use strict';
```
Use same casing for naming migration files

underscope commented on server/shared/database/migrations/20201113162457-add_element_id_to_comment.js:
```diff
@@ -0,0 +1,12 @@
+  allowNull: true
+});
+
+exports.down = qi => qi.removeColumn(TABLE_NAME, 'content_element_id');
```
Use COLUMN_NAME

underscope commented on server/shared/database/migrations/20201113162457-add_element_id_to_comment.js:
```diff
@@ -0,0 +1,12 @@
+exports.up = (qi, { INTEGER }) => qi.addColumn(TABLE_NAME, COLUMN_NAME, {
+  type: INTEGER,
+  references: { model: 'content_element', key: 'id' },
+  allowNull: true
```
Default

underscope commented on client/store/plugins/vuex-persist.js:
```diff
@@ -13,7 +13,7 @@ export default new VuexPersistence({
     repository: {
       comments: {
-        seenByActivity: state.repository.comments.seenByActivity
+        seenBy: state.repository.comments.seenBy
```
Migration / backwards support  needed

underscope commented on client/components/common/tce-core/Discussion/index.vue:
```diff
@@ -41,45 +43,42 @@
 import orderBy from 'lodash/orderBy';
 import TextEditor from './TextEditor';
 
+const SHOWN_COMMENTS_LIMIT = 5;
```
Not reused and named within computed, please omit or consider moving it to the prop.

underscope commented on client/components/editor/ContentElement.vue:
```diff
@@ -1,5 +1,30 @@
 <template>
-  <div>
+  <div class="content-element-container">
+    <v-menu
```
Needs to be within `tce-core/ContentElement`, this one is not always used and needs to be merged. Comments section should be extracted to a separate component.

underscope commented on client/components/editor/ContentElement.vue:
```diff
@@ -1,5 +1,30 @@
+        <v-tooltip open-delay="800" left>
+          <template v-slot:activator="{ on: tooltip }">
+            <v-btn v-on="{ ...menu, ...tooltip }" small icon fab>
+              <v-badge
```
It can just be a small fab?

underscope commented on client/components/editor/ContentElement.vue:
```diff
@@ -29,12 +55,21 @@ export default {
+  }),
+  computed: {
+    ...mapChannels({ editorChannel: 'editor' }),
+    ...mapGetters('repository/comments', ['getUnseenComments']),
```
Can't work like this due to prev comment `tce-core/ContentElement` being decoupled.

underscope commented on client/store/modules/repository/comments/index.js:
```diff
@@ -4,7 +4,10 @@ import * as mutations from './mutations';
 const state = {
   items: {},
-  seenByActivity: {},
+  seenBy: {
```
Since it is now nested, you can omit `By` from name.

underscope commented on client/components/repository/common/Discussion.vue:
```diff
@@ -0,0 +1,74 @@
+<template>
```
Consider having separate wrappers for editor and elements (since their style, behaviour and mappings might be different)

lukasaric committed:
Solved issues partially

- Migration cleanup
- Vuex seen state naming correction
- Set shown comments limit as prop

lukasaric committed:
Merge branch 'develop' into feature/ce-comments

lukasaric committed:
Used discussion wrappers inside tce-ce

lukasaric committed:
Removed leftover props

lukasaric committed:
Set editor channel on CE for activity setup

lukasaric committed:
Class naming polish

lukasaric committed:
Component naming polish :surfer:

lukasaric committed:
Renamed tce CE discussion display flag

lukasaric committed:
Migrate activity seen in vuex persist

lukasaric committed:
Hide CE discussion in revision items

lukasaric committed:
Cleanup

lukasaric committed:
Extracted common options & actions in mixin

lukasaric committed:
Removed unnecessary activity prop from revision

lukasaric committed:
:wrench:

lukasaric committed:
Prepared getters for future sidebar CE comments

lukasaric committed:
Merge branch 'develop' into feature/ce-comments

lukasaric committed:
UI tweaks :art:

lukasaric committed:
Vuex persist migrate seen data cleanup

lukasaric committed:
Added seen backwards compatibility

lukasaric committed:
Vuex persist cleanup :wrench:

lukasaric committed:
:lipstick:

lukasaric committed:
Seen migration more adjustments...

lukasaric committed:
Cleanup :lipstick:

lukasaric committed:
Removed unecessary fetch observe mutation

underscope commented on client/store/modules/repository/comments/mutations.js:
```diff
@@ -8,10 +8,11 @@ import {
-  state.seenByActivity = {
-    ...state.seenByActivity,
-    [activityUid]: lastCommentAt
+const markSeenComments = (state, { activityUid, ceUid, lastCommentAt }) => {
```
Consider renaming `ceUid` to `elementUid` (since we use `elementId` elswhere inlucing the router and it is more consistent)

underscope commented on client/components/common/tce-core/ContentElement.vue:
```diff
@@ -32,20 +32,18 @@ export default {
     frame: { type: Boolean, default: true },
-    dense: { type: Boolean, default: false }
+    dense: { type: Boolean, default: false },
+    displayDiscussion: { type: Boolean, default: true }
```
Consider setting it false by default.

underscope commented on client/components/common/tce-core/ContentElement.vue:
```diff
@@ -62,6 +60,7 @@ export default {
     this.elementBus.on('save:meta', meta => this.$emit('save:meta', meta));
     this.elementBus.on('delete', () => this.$emit('delete'));
     // Editor listeners
+    this.editorChannel.on('activity:set', activity => (this.activity = activity));
```
Why does the element discussion need activity context at this point? 🤔 (Context should be applied within editor)

underscope commented on client/components/common/tce-core/ContentElement.vue:
```diff
@@ -3,24 +3,24 @@
 
 <script>
 import { getComponentName, getElementId } from './utils';
+import Discussion from '@/components/repository/common/Comments/ContentElementDiscussion';
```
Root for this shouldn't be repository (`tce-core` elements should only import `tce-core` elements) 😉

underscope commented on client/components/editor/Sidebar/index.vue:
```diff
@@ -50,9 +50,9 @@
-import ActivityDiscussion from './Discussion';
 import ActivityNavigation from './Navigation';
 import debounce from 'lodash/debounce';
+import EditorDiscussion from '@/components/repository/common/Comments/EditorDiscussion';
```
EditorDiscussion should be within /editor folder rather than common 😉

underscope commented on client/components/repository/common/Sidebar/Body.vue:
```diff
@@ -58,15 +58,15 @@
 import { mapActions, mapGetters } from 'vuex';
 import ActivityTasks from './Tasks';
-import Discussion from './Discussion';
+import Discussion from '../Comments/EditorDiscussion';
```
If a component is used here it should be renamed

underscope commented on client/components/repository/common/Comments/ContentElementDiscussion.vue:
```diff
@@ -0,0 +1,106 @@
+
+<script>
+import Discussion from 'tce-core/Discussion';
+import discussionMixin from './common';
```
By doing this you mapped the store to this component and then used it in `tce-core` (implicitly coupled `tce-core` component with the store)

underscope commented on client/components/repository/common/Comments/EditorDiscussion.vue:
```diff
@@ -0,0 +1,44 @@
+<template>
```
See prev comment regarding usage and name of this file (This was previously and still used for displaying ActivityDiscussions both in Editor and Outline view)

underscope commented on client/store/plugins/vuex-persist.js:
```diff
@@ -13,10 +13,17 @@ export default new VuexPersistence({
   filter: mutation => OBSERVED_MUTATIONS.includes(mutation.type)
 }).plugin;
+
+function migrateSeenState(state) {
```
Consider rather checking upon boot if there is a key in local storage, change its location, and delete it.

underscope commented on client/store/modules/repository/comments/getters.js:
```diff
@@ -1,17 +1,24 @@
+  return filter(items, it => setConditions(it, options));
 };
+
+function setConditions(it, { user, lastSeen, activity, ce }) {
```
Consider eliminating short names like ce here 😉 F definition shouldn't have parameter names like it.

underscope replied:
The F name is unclear.

underscope commented on client/store/modules/repository/comments/getters.js:
```diff
@@ -1,17 +1,24 @@
-    new Date(it.createdAt).getTime() > lastSeen);
+export const getUnseenComments = (state, _, { auth }) => (activity, ce = null) => {
+  const { seen, items } = state;
+  const lastSeen = seen.contentElement[ce?.uid] || seen.activity[activity.uid] || 0;
```
If you want to have a common method then first detect `type` and based on that call two separate handlers. Content elements have `activityId` on them which can be used for this purpose (or just send entityType and entity). With that being said, there might not be a case/need to do so.

underscope commented on client/components/repository/common/Comments/common.js:
```diff
@@ -0,0 +1,39 @@
+    ...mapGetters('repository/comments', ['getComments']),
+    user: vm => vm.$getCurrentUser(),
+    params: vm => extractParams(vm),
+    comments: vm => vm.getComments(vm.params),
```
Comments retrieval can't work this way for content elements (decoupled).

underscope commented on client/components/repository/common/Comments/common.js:
```diff
@@ -0,0 +1,39 @@
+import { mapActions, mapGetters, mapMutations } from 'vuex';
```
In general, there is no need for this mixin since functionality can be only used for ActivityDiscussion component 😉

underscope commented on client/components/repository/common/Comments/ContentElementDiscussion.vue:
```diff
@@ -0,0 +1,106 @@
+<template>
```
Should be a `tce-core` component

underscope commented on client/components/repository/common/Comments/ContentElementDiscussion.vue:
```diff
@@ -0,0 +1,106 @@
+<template>
+  <div class="content-element-discussion">
```
Do we need extra wrapper

underscope commented on client/components/repository/common/Comments/ContentElementDiscussion.vue:
```diff
@@ -0,0 +1,106 @@
+</script>
+
+<style lang="scss" scoped>
+$white: #fff;
```
Should not name variable by color name => `$background-color`

underscope commented on client/components/repository/common/Comments/ContentElementDiscussion.vue:
```diff
@@ -0,0 +1,106 @@
+  }
+}
+
+::v-deep {
```
Increase specificity when using deep

underscope reviewed (commented):
In addition to the PR comments, don't forget to implement email notifications for content elements 😉

lukasaric committed:
Addressed issues partially :scroll:

- Moved CE discussion within tce-core folder
- Extract activity-discussion wrapper in common
- Removed unecessary mixin
- Styles cleanup :lipstick:

lukasaric committed:
Get activity from store on CE discussion

lukasaric committed:
Get activity within editor module

lukasaric committed:
Migrate store state within element bus :rocket:

lukasaric committed:
Unseen comments minor fix & events init cleanup

lukasaric committed:
Removed leftover editor channel import

lukasaric committed:
Getters F & vars naming polish

lukasaric committed:
Extracted unseen comments in separate functions

lukasaric committed:
Decoupled unseen comments getters

lukasaric committed:
Extract element discussion proxy as mixin

- Handled  store state migration within element bus in proxy mixin above

lukasaric committed:
Handled comment email notifications for CE

lukasaric committed:
Element discussion mixin cleanup :lipstick:

lukasaric committed:
Element discussion & vuex persist cleanup

lukasaric committed:
Migrate seen state on boot :rocket:

lukasaric committed:
Seen state :wrench:

lukasaric committed:
Cleanup

lukasaric committed:
Merge branch 'develop' into feature/ce-comments

lukasaric committed:
Merge branch 'develop' into feature/ce-comments

lukasaric committed:
Merge branch 'develop' into feature/ce-comments

lukasaric committed:
Renamed editor channel to editor bus

lukasaric committed:
Merge branch 'develop' into feature/ce-comments

underscope commented on client/components/common/tce-core/ContentElement.vue:
```diff
@@ -15,24 +16,14 @@
-      :is-dragged="isDragged"
-      :is-disabled="isDisabled"
-      :dense="dense" />
-    <v-progress-linear
```
This component is wrongly eliminated upon merge

underscope commented on client/components/common/tce-core/ContentElement.vue:
```diff
@@ -7,6 +7,7 @@
       frame
     }"
     class="content-element">
+    <discussion v-if="displayDiscussion" :element="element" />
```
Consider using show/hide rather than display as a prefix :bowtie:

underscope commented on client/components/common/tce-core/ContentElement.vue:
```diff
@@ -45,7 +36,8 @@ export default {
     frame: { type: Boolean, default: true },
-    dense: { type: Boolean, default: false }
+    dense: { type: Boolean, default: false },
+    displayDiscussion: { type: Boolean, default: false }
```
Is there a place when this false and element is not disabled at the same time?

lukasaric replied:
No there is not. But I think I should improve condition for showing discussion. If element is `disabled`, there's pointless to display discussion within CE. Your thoughts?

underscope commented on client/components/editor/Sidebar/index.vue:
```diff
@@ -132,4 +129,10 @@ export default {
   }
 }
+
+::v-deep .activity-discussion {
```
Please scope it with parent/wrapper class when targeting with ::v-deep

lukasaric committed:
Handled comment data migration within editor

lukasaric committed:
Coverted computed data to collection

lukasaric committed:
Cleanup :lipstick:

underscope commented on client/components/repository/common/ActivityDiscussion.vue:
```diff
@@ -0,0 +1,68 @@
+
+export default {
+  name: 'activity-discussion',
+  inject: ['$getCurrentUser'],
```
For bound components, you can get it from the store (since already mapping a lot of stuff)

underscope commented on client/components/repository/common/ActivityDiscussion.vue:
```diff
@@ -0,0 +1,68 @@
+  inject: ['$getCurrentUser'],
+  props: {
+    activity: { type: Object, required: true },
+    isVisible: { type: Boolean, default: false },
```
I wonder if we can solve this with v-intersect

lukasaric replied:
I should explore it on `vuetify` docs. Then I'll see what I can do...

underscope commented on client/components/repository/common/ActivityDiscussion.vue:
```diff
@@ -0,0 +1,68 @@
+    ...mapGetters('repository/comments', ['getComments']),
+    user: vm => vm.$getCurrentUser(),
+    comments: vm => vm.getComments({ activityId: vm.activity.id, contentElementId: null }),
+    lastCommentAt: vm => new Date(get(vm.comments[0], 'createdAt', 0)).getTime()
```
This component should probably sort it (again) since we cannot rely upon outside ordering not being changed

underscope commented on client/components/editor/index.vue:
```diff
@@ -39,20 +48,72 @@ export default {
   computed: {
-    ...mapGetters('repository/userTracking', ['getActiveUsers']),
-    ...mapGetters('repository', ['repository', 'outlineActivities']),
+    ...mapChannels({ editorBus: 'editor' }),
```
This is a root view with a lot of stuff going on and it shouldn't be polluted with so much of comment-related logic. Extract to a mixin.

underscope commented on client/components/editor/index.vue:
```diff
@@ -19,14 +19,23 @@
+import transform from 'lodash/transform';
 import withUserTracking from 'components/common/mixins/userTracking';
 
+const COMMENT_EVENTS = [
```
Since you'll need these in element as-well, consider extracting them into tce-core.

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,123 @@
+    min-width="300"
+    attach offset-y left>
+    <template v-slot:activator="{ on: menu }">
+      <v-tooltip open-delay="800" left>
```
We should probably create component with menu and tooltip for reuse (we are using it on couple of places and it ends up with 🎄 )

lukasaric replied:
I agree. I think this should be handled in a separate PR. :bowtie:

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,123 @@
+  inject: ['$getCurrentUser'],
+  props: {
+    element: { type: Object, required: true },
+    showHeading: { type: Boolean, default: true }
```
This one will be always the same so it might make sense to omit it.

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,123 @@
+    unseenComments: [],
+    lastCommentAt: 0,
+    unseenCommentCount: 0,
+    showDiscussion: false
```
can be just visible or isVisible since it is referring to the whole component. Consider listing it first.

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,123 @@
+  },
+  data: () => ({
+    comments: [],
+    unseenComments: [],
```
You can use a single array and have unseen determined by last seen date

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,123 @@
+    comments: [],
+    unseenComments: [],
+    lastCommentAt: 0,
+    unseenCommentCount: 0,
```
Should be just a length of previous

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,123 @@
+  }),
+  computed: {
+    ...mapChannels({ editorBus: 'editor' }),
+    elementId: vm => vm.element.id,
```
this is not how element id resolution works, see content element

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,123 @@
+  computed: {
+    ...mapChannels({ editorBus: 'editor' }),
+    elementId: vm => vm.element.id,
+    elementBus: vm => vm.$radio.channel(`element:${vm.elementId}`),
```
Element bus is already provided from parent component, just inject it.

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,123 @@
+    ...mapChannels({ editorBus: 'editor' }),
+    elementId: vm => vm.element.id,
+    elementBus: vm => vm.$radio.channel(`element:${vm.elementId}`),
+    user: vm => vm.$getCurrentUser()
```
Can be passed via prop.

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,123 @@
+    }
+  },
+  created() {
+    this.elementBus.on(EVENTS.INIT, data => Object.assign(this, data));
```
⚠️ 🚨  Other uses can post comments over time. You cannot rely on since INIT.

underscope commented on client/components/editor/index.vue:
```diff
@@ -39,20 +48,72 @@ export default {
+    async initializeComments() {
+      await this.fetchComments({ activityId: this.activityId });
+      const { elements, editorBus } = this;
+      setTimeout(() => forEach(elements, it => this.emitCommentsData(it.id)), 1000);
```
You cannot initialize it once (see comment within ElementDiscussion). Since elements are passed to all containers within ContainerList component now (I unified same approach on the latest develop), you can now just create a computed which will have comments attached so the state will always be properly hydrated.

underscope commented on server/shared/mail/index.js:
```diff
@@ -22,10 +22,14 @@ const send = promisify(server.send.bind(server));
+  urlJoin(origin, '/#/repository', `${repositoryId}?activityId=${activityId}`);
+const elementUrl = ({ repositoryId, activityId, elementUid }) => {
+  const query = `${activityId}?elementId=${elementUid}`;
+  return urlJoin(origin, '/#/repository', String(repositoryId), 'editor', query);
```
Interpolate with editor to make it a bit more compact

lukasaric committed:
Partially addressed issues :surfer:

- Styles cleanup
-  Passed hydrated store within containers props
- Naming polish

lukasaric committed:
Omit show heading from props

lukasaric committed:
Events cleanup

lukasaric committed:
Removed duplicated getter

lukasaric committed:
Remove comment deletion method wrapper

underscope commented on client/components/common/tce-core/utils.js:
```diff
@@ -32,6 +32,12 @@ export function getElementId(element) {
   return element && (element.uid || element.id);
 }
 
+export const getCommentEvents = () => ({
```
Please replace with standard enum like const. Instead of utils create Events dir and expose CommentEvent/DiscussionEvent

underscope commented on client/components/editor/ActivityContent/ContainerList.vue:
```diff
@@ -54,7 +54,8 @@ export default {
   props: {
-    containerGroup: { type: Array, default() { return []; } },
+    containerGroup: { type: Array, default: () => ({}) },
+    commentsWithinElements: { type: Object, required: true },
```
Since there will be other things included use something more generic like processedElements

underscope commented on client/components/common/mixins/elementDiscussion.js:
```diff
@@ -0,0 +1,55 @@
+    ...mapChannels({ editorBus: 'editor' }),
+    ...mapState('repository/contentElements', { elements: 'items' }),
+    ...mapGetters('repository/comments', ['getUnseenElementComments', 'getComments']),
+    commentsWithinElements() {
```
Expose processor for adding them rather than computed.  Use it then in ActivityContent (since there will be other ones as well)

underscope commented on client/components/common/mixins/elementDiscussion.js:
```diff
@@ -0,0 +1,55 @@
+
+const { SAVE, REMOVE, SET_LAST_SEEN } = getCommentEvents();
+
+const COMMENT_EVENTS = [
```
Consider extracting only comment event handling into separate mixin

underscope commented on client/components/editor/ActivityContent/index.vue:
```diff
@@ -11,7 +11,9 @@
           v-for="(containerGroup, type) in rootContainerGroups"
           :key="type"
           :container-group="containerGroup"
+          :comments-within-elements="commentsWithinElements"
```
See mixin comment

underscope commented on client/components/editor/ActivityContent/index.vue:
```diff
@@ -11,7 +11,9 @@
           :container-group="containerGroup"
+          :comments-within-elements="commentsWithinElements"
           :parent-id="activity.id"
+          :activity="activity"
```
Probably does not make sense combined with parent-id

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,116 @@
+    id: { type: Number, required: true },
+    uid: { type: String, required: true },
+    comments: { type: Array, required: true },
+    unseenComments: { type: Array, required: true },
```
Lets discuss separating these two

lukasaric replied:
I removed CE unseen processor from `comment` getter. Then I passed `lastSeen` state per element and handle unseen comments within this `tce-core` wrapper as computed. As for the `unseenCommentCount`, i need that, because I need short delay (handled in watcher) for displaying unseen comment notification. Tomorrow I will do the cleanup, and see if there's more improvements here. After that let me now if there's some more of your thoughts about this approach...

underscope replied:
Can't the delay be created by triggering a delayed event 🤔 ?

lukasaric committed:
Comments handling improvements :surfer:

- Extract element processor as getter
- Naming cleanup
- Create tce-core events dir
- Exposed comment events in dir above
- Removed leftover activity prop
- Activity discussion comments order

lukasaric committed:
Handled CE unseen inside tce wrapper

lukasaric committed:
Process last seen in processed elements

underscope commented on client/components/common/mixins/elementDiscussion.js:
```diff
@@ -0,0 +1,39 @@
+    }
+  },
+  async mounted() {
+    await this.fetchComments({ activityId: this.activity.id });
```
This one as well; 😉 And then rename it to something like commentEventListeners

underscope commented on client/components/common/mixins/elementDiscussion.js:
```diff
@@ -0,0 +1,39 @@
+
```
Empty line leftover

lukasaric replied:
🙈

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,123 @@
+    uid: { type: String, required: true },
+    comments: { type: Array, required: true },
+    lastSeen: { type: Number, required: true },
+    lastCommentAt: { type: Number, required: true },
```
Isn't this a computed based on comments array? 🤔

lukasaric replied:
Yes. I was planning to move this here within wrapper as computed.

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,123 @@
+  },
+  data: () => ({
+    isVisible: false,
+    unseenCommentCount: 0
```
unseenComments.length? 🤔

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,123 @@
+    },
+    setLastSeen(timeout) {
+      const { uid: elementUid, lastCommentAt, events } = this;
+      const options = { timeout, elementUid, lastCommentAt };
```
order by importance (elementId first)

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,123 @@
+    position: relative;
+  }
+
+  ::v-deep .v-badge {
```
Not sure if this is the intention of the component when used within btn. Maybe we should use an avatar within btn with a number to avoid this.

underscope commented on client/components/common/tce-core/Events/discussionEvent.js:
```diff
@@ -0,0 +1,5 @@
+export default {
```
Casing of the file should be PascalCase

underscope commented on client/components/editor/ActivityContent/index.vue:
```diff
@@ -56,9 +59,10 @@ export default {
     ...mapGetters('repository', ['activities']),
     ...mapGetters('editor', ['collaboratorSelections']),
-    ...mapChannels({ editorChannel: 'editor' }),
+    ...mapGetters('repository/contentElements', ['processedElements']),
```
If it is going to be used only in a single place, it might not make sense to put it in a store (it is better to contexualize it to make it more readable)

underscope commented on client/store/modules/repository/content-elements/getters.js:
```diff
@@ -1 +1,16 @@
+
+export const processedElements = (_, { elements }, rootState, rootGetters) => {
+  const { id: activityId } = rootGetters['editor/activity'];
+  const getComments = rootGetters['repository/comments/getComments'];
```
Fetch comments at the same time as content elements (currently in a mixin).

underscope commented on client/store/modules/repository/content-elements/getters.js:
```diff
@@ -1 +1,16 @@
+  const { seen } = rootState.repository.comments;
+  return transform(elements, (acc, it) => {
+    const comments = getComments({ activityId, contentElementId: it.id });
+    const lastCommentAt = new Date(get(comments[0], 'createdAt', 0)).getTime();
```
Why do we need it precalculated (has performance implications, processing will be done on each view of unseen). Should be scoped if possible to the respective component.

lukasaric committed:
CE comment enhancement :rocket:

- Moved processedElements within component
- Renamed mixin & handled events only in it
- Cleanup

lukasaric committed:
Replaced badge with v-avatar

lukasaric committed:
Cleanup :lipstick:

lukasaric committed:
Comment author id condition fix

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,114 @@
+            <v-icon v-else color="primary" class="pr-1">mdi-forum-outline</v-icon>
+          </v-btn>
+        </template>
+        <span>Discussion</span>
```
Let's improve the label by setting it to:

- `Post a comment` if there are no comments
- `View comments` if there are comments.

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,114 @@
+    <template v-slot:activator="{ on: menu }">
+      <v-tooltip open-delay="800" left>
+        <template v-slot:activator="{ on: tooltip }">
+          <v-btn v-on="{ ...menu, ...tooltip }" icon>
```
Please apply changes and fix issues from PM

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,114 @@
+  },
+  data: () => ({
+    isVisible: false,
+    unseenCommentCount: 0
```
Leftover?

lukasaric committed:
Improved tooltip message & cleanup

lukasaric committed:
Propagate case sensitive event file name

lukasaric committed:
UI tweaks & improvements :art:

lukasaric committed:
Implemented `v-intersect`

lukasaric committed:
Set discussion ref on container instead of editor

underscope commented on client/components/common/tce-core/Events/DiscussionEvent.js:
```diff
@@ -0,0 +1,5 @@
+export default {
+  SAVE: 'comment:save',
+  REMOVE: 'comment:remove',
+  SET_LAST_SEEN: 'comment:set-last-seen'
```
namespace:camelCaseEventName

underscope commented on client/components/repository/common/ActivityDiscussion.vue:
```diff
@@ -0,0 +1,74 @@
+      const payload = { activityUid: activity.uid, lastCommentAt };
+      setTimeout(() => this.markSeenComments(payload), timeout);
+    },
+    onIntersect(_, __, isIntersected) {
```
consider signaling what params are in case of future changes i.e. `_entries`

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,129 @@
+            <v-avatar v-if="unseenComments.length" size="16" color="secondary">
+              {{ unseenComments.length }}
+            </v-avatar>
+            <v-icon v-else :color="options.color">{{ options.icon }}</v-icon>
```
Options sound a bit too generic 🤔 ?  You can further improve it (if you are going to have options like set) by eliminating avatar and setting a class with the background color (v-btn icon + class like pink will do the same thing) and then have something like `commentActivator` computed which will return label, icon, text and class names.

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,129 @@
+      @remove="editorBus.emit(events.REMOVE, $event)"
+      v-bind="{ comments, user }"
+      show-heading
+      class="pa-4" />
```
Try to make it a bit more compact

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -0,0 +1,129 @@
+$background-color: #fff;
+
+.v-menu__content {
+  background: $background-color;
```
No need to set these things over here (same below for avatar and icon)

underscope commented on client/components/common/tce-core/Discussion/index.vue:
```diff
@@ -52,45 +52,40 @@ export default {
-      this.$nextTick(() => this.editor.scrollIntoView());
+      this.$emit('save', payload);
+      // Keep discussion container inside viewport.
+      this.$nextTick(() => this.discussion.scrollIntoView({ behavior: 'smooth' }));
```
This broke now behaviour within activity sidebar (outline view)

lukasaric committed:
Cleanup

- Styles & naming convention polish
- Scroll & sidebar bar show comments issue fix

lukasaric committed:
Improved activator options

lukasaric committed:
Minor padding adjustments

underscope commented on client/components/common/tce-core/Events/DiscussionEvent.js:
```diff
@@ -0,0 +1,5 @@
+export default {
+  save: 'comment:save',
+  remove: 'comment:remove',
+  setLastSeen: 'comment:set-last-seen'
```
Forgot to change event name casing ? kebab-case. More often we define namespace as the second part. Sorry, I forgot to note on previous comment.

lukasaric committed:
Event naming cleanup :lipstick:

lukasaric committed:
Merge branch 'develop' into feature/ce-comments

underscope committed:
Show add comment btn only on hover / focus

- Added element actions toolbar
- Passed isHovered to ContentElement

underscope committed:
Fix input area sizing based on comment
